File: J9Ch/src/S_2_Data_Structure_I/Required_12/python_cpp/_589_Connecting_Graph_Medium.cpp



class ConnectingGraph {
private:
    vector<int> father;
public:
    int find(int x) {
        if (father[x] == 0)
            return x;

        return father[x] = find(father[x]);
    }

    ConnectingGraph(int n) {
        // initialize your data structure here.
        father.resize(n + 1);
        for (int i = 1; i <= n; ++i)
            father[i] = 0;
    }

    void  connect(int a, int b) {
        // Write your code here
        int root_a = find(a);
        int root_b = find(b);
        if (root_a != root_b)
            father[root_a] = root_b;
    }

    bool query(int a, int b) {
        // Write your code here
        int root_a = find(a);
        int root_b = find(b);
        return root_a == root_b;
    }
};

File: J9Ch/src/S_2_DS_I/Required_12/python_cpp/_589_Connecting_Graph_Medium.py


class ConnectingGraph:

    # @param {int} n
    def __init__(self, n):
        # initialize your data structure here.
        self.father = [0 for _ in xrange(n + 1)]

    def find(self, x):
        if self.father[x] == 0:
            return x
        self.father[x] = self.find(self.father[x])
        return self.father[x]

    # @param {int} a, b
    # return nothing
    def connect(self, a, b):
        # Write your code here
        root_a = self.find(a)
        root_b = self.find(b)
        if root_a != root_b:
            self.father[root_a] = root_b

    # @param {int} a, b
    # return {boolean} true if they are connected or false
    def query(self, a, b):
        # Write your code here
        root_a = self.find(a)
        root_b = self.find(b)
        return root_a == root_b
